### PR TITLE
`<MultiSelect/>` (fix)-  Pressing Tab should focusing out of component

### PR DESCRIPTION
**Pull request by mosheKerbel**
PR solves this issue:
https://github.com/wix-private/wsr-issues/issues/263

Still working on e2e tests for
-  inputWithOptions

### CODE REVIEW

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 1]
```diff
@@ -1,61 +1,143 @@
...
+        storyName: storySettings.storyName,
+        dataHook: storySettings.dataHook,
+        testName,
+      });
+      await browser.get(testStoryUrl);
+    };
 
-    expect(await driver.getDropdownItemsCount()).toEqual(30);
-  });
+    beforeAll(async () => {
```
Comment (shlomitc) [inline src/DropdownLayout/DropdownLayout.e2e.js]: remove `beforeAll`

File: src/InputWithOptions/InputWithOptions.js [hunk 2]
```diff
@@ -232,7 +233,7 @@ class InputWithOptions extends WixComponent {
     return this.getManualSubmitKeys().includes(key);
   }
 
-  _onManuallyInput(inputValue = '') {
+  _onManuallyInput(inputValue = '', event) {
```
Comment (shlomitc) [inline src/InputWithOptions/InputWithOptions.js]: this is redundant

File: src/InputWithOptions/InputWithOptions.js [hunk 3]
```diff
@@ -315,6 +316,7 @@ class InputWithOptions extends WixComponent {
     if (this.props.onBlur) {
       this.props.onBlur(e);
     }
+    this.hideOptions();
```
Comment (shlomitc) [inline src/InputWithOptions/InputWithOptions.js]: why this change was introduced? if this is the correct behavior, it should be added with a JSDOM test

File: src/InputWithOptions/docs/storySettings.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { Category } from '../../../stories/storiesHierarchy';
+
+export const storySettings = {
+  storyName: 'InputWithOptions',
+  indexCategory: Category.SELECTION,
+  dataHook: 'story-input-with-options',
+};
+
+export const testStories = {
+  e2e: 'e2e'
```
Comment (shlomitc) [inline src/InputWithOptions/docs/storySettings.js]: rename

File: src/InputWithOptions/InputWithOptions.js [hunk 5]
```diff
@@ -160,15 +160,16 @@ class InputWithOptions extends WixComponent {
...
       </div>
     );
   }
 
   render() {
-    const { dropDirectionUp } = this.props;
+    const { dropDirectionUp, dataHook } = this.props;
     return (
-      <div>
+      <div data-hook={dataHook} >
```
Comment (hepiyellow) [inline src/InputWithOptions/InputWithOptions.js]: `InputWithOptions` extends `WixComponent` which applies the `dataHook` prop onto the root element at `componentDidMount()`.
Refactoring this is out of this tasks scope.

File: src/DropdownLayout/DropdownLayout.e2e.js (hunk 1 above)
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/DropdownLayout/docs/storySettings.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { Category } from '../../../stories/storiesHierarchy';
+
+export const storySettings = {
+  storyName: 'DropdownLayout',
+  indexCategory: Category.PICKERS_AND_SELECTORS,
+  dataHook: 'story-dropdown-layout',
+};
+
+export const testStories = {
+  basic: 'Basic'
```
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/DropdownLayout/docs/storySettings.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 7]
```diff
@@ -1,61 +1,143 @@
...
 
-  eyes.it('should render items properly', async () => {
-    expect(await driver.getDropdownItem(0)).toEqual('options 0');
+      expect(await driver.loaderExists()).toBeTruthy();
+    });
   });
 
-  eyes.it('should add more items from server when scrolling down', async () => {
-    expect(await driver.getDropdownItemsCount()).toEqual(15);
+  describe('Part B', () => {
```
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 8]
```diff
@@ -1,61 +1,143 @@
...
-  eyes.it('should show loader', async () => {
-    await driver.scrollToElement(14);
+    const pressTab = () =>
+      browser
+        .actions()
+        .sendKeys(protractor.Key.TAB)
+        .perform();
 
-    expect(await driver.loaderExists()).toBeTruthy();
+    eyes.it(
```
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 9]
```diff
@@ -1,61 +1,143 @@
...
-  });
+    beforeAll(async () => {
+      beforeEach(async () => {
+        await navigateToTestUrl(testStories.basic);
+      });
+    });
+
+    beforeEach(async () => {
+      driver = dropdownLayoutTestkitFactory({
+        dataHook: 'many-options-dropdown-layout',
```
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 10]
```diff
@@ -1,61 +1,143 @@
...
-    await driver.scrollToElement(14);
+    const pressTab = () =>
+      browser
+        .actions()
+        .sendKeys(protractor.Key.TAB)
+        .perform();
 
-    expect(await driver.loaderExists()).toBeTruthy();
+    eyes.it(
+      'should move out focus of dropdown only after 2 tab press when selecting an item',
```
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 11]
```diff
@@ -1,61 +1,143 @@
...
+    eyes.it(
+      'should move out focus of dropdown when pressing tab without any selection',
+      async () => {
+        const firstElement = await driver.getElementByDataHook(
+          'input-for-initial-focus',
+        );
+        pressTab();
+        expect(await driver.isFocused(firstElement)).toEqual(true);
+        pressTab();
+        const real = await driver.getElementByDataHook(
```
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/InputWithOptions/InputWithOptions.e2e.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
+import { inputWithOptionsTestkitFactory } from '../../testkit/protractor';
```
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/InputWithOptions/InputWithOptions.e2e.js]: done

File: src/InputWithOptions/InputWithOptions.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/InputWithOptions/InputWithOptions.js]: done

File: src/InputWithOptions/InputWithOptions.js (hunk 2 above)
Comment (mosheKerbel) [reply src/InputWithOptions/InputWithOptions.js]: done

File: src/InputWithOptions/InputWithOptions.js (hunk 3 above)
Comment (mosheKerbel) [reply src/InputWithOptions/InputWithOptions.js]: it is mandatory for the foucs movment upon tab swtiching
it did not behave correctly until now.
i had to modify the expected test result here:
https://github.com/wix/wix-style-react/blob/d9dbf5565260aecf423635daa0a19c1872d84005/src/MultiSelect/MultiSelect.spec.js#L504

File: src/InputWithOptions/docs/storySettings.js (hunk 4 above)
Comment (mosheKerbel) [reply src/InputWithOptions/docs/storySettings.js]: done

File: src/InputWithOptions/InputWithOptions.js (hunk 5 above)
Comment (mosheKerbel) [reply src/InputWithOptions/InputWithOptions.js]: reverted, i see that it really works using WixComponent
not sure what happened before.

File: src/InputWithOptions/InputWithOptions.js [hunk 13]
```diff
@@ -160,13 +160,14 @@ class InputWithOptions extends WixComponent {
...
           isComposing={this.state.isComposing}
+          tabIndex={-1}
         />
       </div>
     );
   }
 
   render() {
-    const { dropDirectionUp } = this.props;
+    const { dropDirectionUp, dataHook } = this.props;
```
Comment (hepiyellow) [inline src/InputWithOptions/InputWithOptions.js]: Unused `dataHook`

File: src/InputWithOptions/docs/tests/TestTabSwitches.js [hunk 14]
```diff
@@ -0,0 +1,87 @@
+/* eslint-disable no-console */
```
Comment (hepiyellow) [inline src/InputWithOptions/docs/tests/TestTabSwitches.js]: I think you meant to put `test`folder under `src/InputWithOptions`, not under `docs`

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 15]
```diff
@@ -1,61 +1,129 @@
...
+
+    const pressTab = () =>
+      browser
+        .actions()
+        .sendKeys(protractor.Key.TAB)
+        .perform();
+
+    async function focusOnDropdownLayout() {
+      const firstElement = $(`[data-hook="input-for-initial-focus"]`);
+      pressTab();
```
Comment (hepiyellow) [inline src/DropdownLayout/DropdownLayout.e2e.js]:
```suggestion
      await pressTab();
```

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 16]
```diff
@@ -1,61 +1,129 @@
...
+      expect(await isFocused(firstElement)).toEqual(true);
 
-  eyes.it('should show loader', async () => {
-    await driver.scrollToElement(14);
+      pressTab();
```
Comment (hepiyellow) [inline src/DropdownLayout/DropdownLayout.e2e.js]:
```suggestion
      await pressTab();
```

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 17]
```diff
@@ -1,61 +1,129 @@
...
+      expect(await driver.isFocused()).toEqual(true);
+    }
 
-    expect(await driver.loaderExists()).toBeTruthy();
+    it('should move out focus of dropdown only after 2 tab press when selecting an item', async () => {
+      await focusOnDropdownLayout();
+
+      await driver.hoverItemById(0);
+      pressTab();
```
Comment (hepiyellow) [inline src/DropdownLayout/DropdownLayout.e2e.js]:
```suggestion
      await pressTab();
```

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 18]
```diff
@@ -1,61 +1,129 @@
...
+      expect(await driver.isFocused()).toEqual(true);
+
+      pressTab();
```
Comment (hepiyellow) [inline src/DropdownLayout/DropdownLayout.e2e.js]:
```suggestion
      await pressTab();
```

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 19]
```diff
@@ -1,61 +1,129 @@
...
+      expect(await driver.isFocused()).toEqual(false);
+    });
+
+    it('should move out focus of dropdown when pressing tab without any selection', async () => {
+      await focusOnDropdownLayout();
+
+      pressTab();
```
Comment (hepiyellow) [inline src/DropdownLayout/DropdownLayout.e2e.js]:
```suggestion
      await pressTab();
```

Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js (hunk 18 above)
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js (hunk 17 above)
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js (hunk 16 above)
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js (hunk 15 above)
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: done

File: src/InputWithOptions/InputWithOptions.js (hunk 13 above)
Comment (mosheKerbel) [reply src/InputWithOptions/InputWithOptions.js]: done

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 20]
```diff
@@ -1,61 +1,143 @@
...
 
-  eyes.it('should add more items from server when scrolling down', async () => {
-    expect(await driver.getDropdownItemsCount()).toEqual(15);
+  describe('Part B', () => {
+    let driver;
 
-    await driver.scrollToElement(14);
-    await waitForFetching();
+    const navigateToTestUrl = async testName => {
+      const testStoryUrl = createTestStoryUrl({
```
[the remark this answers is not in the document]
Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: one is using category and one is using indexCategory
i will have to change it first so it can destruct params correctly

File: src/MultiSelect/MultiSelect.js [hunk 21]
```diff
@@ -103,6 +103,7 @@ class MultiSelect extends InputWithOptions {
     const _value = (value && value.trim()) || (inputValue && inputValue.trim());
 
     this.submitValue(_value);
+    _value && event.preventDefault();
```
[the remark this answers is not in the document]
Comment (milkyfruit) [reply src/MultiSelect/MultiSelect.js]: @hepiyellow 
Spaces alone should be ignored, unless they are in between other chars (e.g. "Name Surname").
So, in this case, it should step out of multiselect and focus to next component.

Behaviour scenarios:
a. Writes "name", on press "Tab accepts the value and keeps focused
b. Writes "name surname", on press Tab accepts the value and keeps focused
c. Writes " ", on press Tab declines the value and focused to next component.

I'll add these notes to Zeplin.

File: src/InputWithOptions/InputWithOptions.js (hunk 3 above)
Comment (mosheKerbel) [reply src/InputWithOptions/InputWithOptions.js]: @shlomitc plz take a look at this latest commit https://github.com/wix/wix-style-react/pull/3080/commits/d6562fec9bda95fd0a11feff1cd2cc99c50d282b
which now handles the options hide properly.

File: src/DropdownLayout/DropdownLayout.e2e.js [hunk 22]
```diff
@@ -1,61 +1,129 @@
...
+import { storySettings, testStories } from './docs/storySettings';
 
 async function waitForFetching() {
   await sleep(700);
 }
 
 describe('DropdownLayout', () => {
-  let driver;
+  // TODO: divided tests to 2 parts - need to migrate Part A to use stroybook tests sections
+  describe('General', () => {
```
Comment (shlomitc) [inline src/DropdownLayout/DropdownLayout.e2e.js]: adding an additional layer for the `eyes.it` tests is problematic since it changes the baseline name
image

This mean that if you accidently broke the test in this PR, it would not identify it since it's considered a new baseline...

I suggest not adding this additinal describe for existing test scenarios

Comment (mosheKerbel) [reply src/DropdownLayout/DropdownLayout.e2e.js]: makes sense 👍 
i've changed it.

File: src/MultiSelect/MultiSelect.e2e.js [hunk 23]
```diff
@@ -29,6 +33,8 @@ describe('MultiSelect', () => {
       eyes.it('should show 2 tags', async () => {
         await driver.click();
         await driver.selectItemById('AL');
+
+        await driver.click();
```
Comment (shlomitc) [inline src/MultiSelect/MultiSelect.e2e.js]: this looks a bit strange, it changes the component behavior.
previously - clicking on the list item, would keep it open, allowing you to selected more items from the list.
currently - it closes right after selecting an item.

File: src/InputWithOptions/InputWithOptions.js [hunk 24]
```diff
@@ -315,6 +316,10 @@ class InputWithOptions extends WixComponent {
     if (this.props.onBlur) {
       this.props.onBlur(e);
     }
+
+    if (this.props.closeOnSelect) {
```
Comment (shlomitc) [inline src/InputWithOptions/InputWithOptions.js]: not sure it should be in the `onBlur`, should probably go to the `onKeyDown`

Comment (hepiyellow) [reply src/InputWithOptions/InputWithOptions.js]: I think @shlomitc  is right.
The use-case I can think of is that the `<DropdownLayout/>` can have a fixed Heade/Footer (It is in the API: `fixedFooter`, `fixedHeader`). And those can have any node, including, for example, a focusable `<Button/>` component. This means that when the user clicks on the `<Button/>` (or navigate to it by keyboard), the input will lose focus, and the options will close.

So yeah, do it in `onKeyDown`

File: src/MultiSelect/MultiSelect.e2e.js (hunk 23 above)
Comment (hepiyellow) [reply src/MultiSelect/MultiSelect.e2e.js]: The options don't close after the first select. So I think the second `driver.click()` is simply redundant.

File: src/DropdownLayout/DropdownLayout.e2e.js (hunk 22 above)
Comment (hepiyellow) [reply src/DropdownLayout/DropdownLayout.e2e.js]: In case you do wish to rename tests, then you can do it in a separate PR which has ONLY the rename change.

File: src/MultiSelect/MultiSelect.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (hepiyellow) [reply src/MultiSelect/MultiSelect.js]: ok. so the behaviour now is what we want.

File: src/DropdownLayout/DropdownLayout.e2e.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (hepiyellow) [reply src/DropdownLayout/DropdownLayout.e2e.js]: I don't see why this `indexCategory` convention was introduced. See my post on slack `wix-private`

File: src/InputWithOptions/InputWithOptions.js (hunk 3 above)
Comment (hepiyellow) [reply src/InputWithOptions/InputWithOptions.js]: This is duplicate of https://github.com/wix/wix-style-react/pull/3080#discussion_r266226213
Resolving

File: src/MultiSelect/MultiSelect.spec.js [hunk 25]
```diff
@@ -531,7 +531,7 @@ describe('MultiSelect', () => {
         driver.pressKey('Tab');
         expect(onSelect).toHaveBeenCalledTimes(1);
         expect(dropdownLayoutDriver.isShown()).toBeTruthy();
-        expect(inputDriver.isFocus()).toBeTruthy();
+        // expect(inputDriver.isFocus()).toBeTruthy(); // Limitation - covered with e2e test
```
Comment (mosheKerbel) [inline src/MultiSelect/MultiSelect.spec.js]: Couldn't manage to make it work with the setTimeout
Need your help guys.

File: src/MultiSelect/MultiSelect.spec.js [hunk 26]
```diff
@@ -501,12 +501,12 @@ describe('MultiSelect', () => {
     });
 
     describe('Keep Options Open', () => {
-      it('should not lose Focus but should close the options when options selected by mouse click', () => {
+      it('should not lose Focus or close the options when options selected by mouse click', () => {
```
Comment (mosheKerbel) [inline src/MultiSelect/MultiSelect.spec.js]: Reverted back to the original test case

File: src/MultiSelect/MultiSelect.js [hunk 27]
```diff
@@ -148,7 +148,9 @@ class MultiSelect extends InputWithOptions {
       onSelect(this.props.options.find(o => o.id === option.id));
     }
 
-    this.input.focus();
+    setTimeout(() => {
+      this.input.focus();
```
Comment (mosheKerbel) [inline src/MultiSelect/MultiSelect.js]: When clicking on an option from the dropdown we get here
We need to set the focus on the input.
Until now it was not focused!
I added the timeout as a suggestion from many similar issues i googled..
